### PR TITLE
Revert most of "Avoid implicit bool conversions"

### PR DESCRIPTION
**Pull request by drwells**
This reverts most of commit bb84662fbdd967da82c140e8d13fddf229dd371d. I kept the bug fix with `numbers::invalid_unsigned_int` and a few other things that improved code clarity. I would prefer if we wrote `v.size() == 0` instead of `!v.size()`, etc.: I think those changes improve readability. All changes to enums were reverted.

The clear winner in the poll from

https://github.com/dealii/dealii/pull/13229#issuecomment-1014767396

is 'return to the previous status quo' so this patch is the agreed upon solution.

### CODE REVIEW

**Comment by drwells** [comment]
This replaces #13229.

**Review by tjhei — APPROVED** [review]
Thank you for doing that. Looks good to me.

> I would prefer if we wrote v.size() == 0 instead of !v.size(), etc.: 

agreed!

File: include/deal.II/meshworker/integration_info.templates.h [hunk 1]
```diff
@@ -191,32 +191,32 @@ namespace MeshWorker
     face_flags |= boundary_flags;
     neighbor_flags |= neighbor_geometry ? boundary_flags : update_default;
 
-    if (cell_selector.has_values())
+    if (cell_selector.has_values() != 0)
```
Comment (drwells) [inline include/deal.II/meshworker/integration_info.templates.h:194]: I just noticed this one - before I looked for `0u`. It looks like old versions of deal.II had the `!= 0` check which converts the left side to an integer when its already a boolean - odd. I'll fix it.

File: include/deal.II/base/geometry_info.h [hunk 2]
```diff
@@ -3879,8 +3879,8 @@ GeometryInfo<2>::min_cell_refinement_case_for_face_refinement(
   AssertIndexRange(face_no, GeometryInfo<dim>::faces_per_cell);
 
   if (face_refinement_case == RefinementCase<dim>::cut_x)
-    return (face_no / 2) != 0u ? RefinementCase<dim>::cut_x :
-                                 RefinementCase<dim>::cut_y;
+    return (face_no / 2) ? RefinementCase<dim>::cut_x :
+                           RefinementCase<dim>::cut_y;
```
Comment (masterleinad) [inline include/deal.II/base/geometry_info.h:3883]: IMHO converting `face_no/2` implicitly to a bool isn't better. Do you think so?
`face_no/2 != 0` essentially means `face_no >=2` I would prefer that.

File: include/deal.II/base/geometry_info.h [hunk 3]
```diff
@@ -3975,8 +3975,7 @@ GeometryInfo<2>::min_cell_refinement_case_for_line_refinement(
   (void)dim;
   AssertIndexRange(line_no, GeometryInfo<dim>::lines_per_cell);
 
-  return (line_no / 2) != 0u ? RefinementCase<2>::cut_x :
-                               RefinementCase<2>::cut_y;
+  return (line_no / 2) ? RefinementCase<2>::cut_x : RefinementCase<2>::cut_y;
```
Comment (masterleinad) [inline include/deal.II/base/geometry_info.h:3978]: same here.

File: include/deal.II/base/patterns.h [hunk 4]
```diff
@@ -2335,8 +2335,7 @@ namespace Patterns
         const auto  string_array = Convert<T>::to_string_internal_2(t, *p);
         std::string str;
         for (unsigned int i = 0; i < string_array.size(); ++i)
-          str +=
-            (i != 0u ? " " + p->get_separator() + " " : "") + string_array[i];
+          str += (i ? " " + p->get_separator() + " " : "") + string_array[i];
```
Comment (masterleinad) [inline include/deal.II/base/patterns.h:2338]: ...and here and a couple other places.

**Review by masterleinad — COMMENTED** [review]
I would prefer avoiding implicitly conversion from `int`s to `bools` where it's not a flag.

**Comment by drwells** [comment]
@masterleinad good point - we only decided on enums anyway. I'll un-revert those.

File: include/deal.II/matrix_free/evaluation_kernels.h [hunk 5]
```diff
@@ -902,7 +902,7 @@ namespace internal
                           n_q_points);
 
                 if ((add_into_values_array == false &&
-                     (integration_flag & EvaluationFlags::values) == 0) &&
+                     (integration_flag & EvaluationFlags::values) == false) &&
```
Comment (masterleinad) [inline include/deal.II/matrix_free/evaluation_kernels.h:905]: Shouldn't this be
```suggestion
                     !(integration_flag & EvaluationFlags::values)) &&
```

File: source/base/table_handler.cc [hunk 6]
```diff
@@ -268,7 +268,7 @@ TableHandler::add_column_to_supercolumn(const std::string &key,
 {
   Assert(columns.count(key), ExcColumnNotExistent(key));
 
-  if (supercolumns.count(superkey) == 0u)
+  if (!supercolumns.count(superkey))
```
Comment (masterleinad) [inline source/base/table_handler.cc:271]: Here, and below I would also prefer the previous version since `count` returns an `int`.

**Comment by drwells** [comment]
Thanks for catching those. I read through the changes again and I think we are now set.

**Comment by masterleinad** [comment]
Almost, we still have
```
The following tests FAILED:
	 58 - simplex/poisson_02.debug (Failed)
	 77 - simplex/step-07.debug (Failed)
```

**Review by masterleinad — APPROVED** [review]

**Pull request merged**